src/poly/multiopen/verifier.rs

```diff
@@ -0,0 +1,246 @@
+}
+
+impl<'a, C: CurveAffine> Proof<C> {
+    /// Verify a multi-opening proof
+    pub fn verify<I, HBase: Hasher<C::Base>, HScalar: Hasher<C::Scalar>>(
+        &self,
+        params: &'a Params<C>,
+        transcript: &mut HBase,
+        transcript_scalar: &mut HScalar,
+        queries: I,
+        msm: MSM<'a, C>,
+    ) -> Result<Guard<'a, C>, Error>
+    where
+        I: IntoIterator<Item = VerifierQuery<'a, C>> + Clone,
+    {
+        // Sample x_4 for compressing openings at the same points together
+        let x_4: C::Scalar = get_challenge_scalar(Challenge(transcript.squeeze().get_lower_128()));
+
+        let (commitment_map, point_sets) = construct_intermediate_sets::<'a, C, I>(queries.clone());
+
+        // Compress the commitments and expected evaluations at x_3 together.
+        // using the challenge x_4
+        let mut q_commitments: Vec<_> = vec![params.empty_msm(); point_sets.len()];
+        let mut q_eval_sets: Vec<Vec<C::Scalar>> = vec![Vec::new(); point_sets.len()];
+        for (set_idx, point_set) in point_sets.iter().enumerate() {
+            q_eval_sets[set_idx] = vec![C::Scalar::zero(); point_set.len()];
+        }
+        {
+            let mut accumulate = |set_idx: usize, new_commitment, evals: Vec<C::Scalar>| {
+                q_commitments[set_idx].scale(x_4);
+                q_commitments[set_idx].add_term(C::Scalar::one(), new_commitment);
+                for (eval_idx, &eval) in evals.iter().enumerate() {
+                    q_eval_sets[set_idx][eval_idx] *= &x_4;
+                    q_eval_sets[set_idx][eval_idx] += &eval;
+                }
+            };
+
+            for (commitment, commitment_data) in commitment_map {
+                accumulate(
+                    commitment_data.set_index,      // set_idx,
+                    *commitment,                    // commitment,
+                    commitment_data.evals.to_vec(), // evals
+                );
+            }
+        }
+
+        // Sample a challenge x_5 for keeping the multi-point quotient
+        // polynomial terms linearly independent.
+        let x_5: C::Scalar = get_challenge_scalar(Challenge(transcript.squeeze().get_lower_128()));
+
+        // Obtain the commitment to the multi-point quotient polynomial f(X).
+        hash_point(transcript, &self.f_commitment).unwrap();
+
+        // Sample a challenge x_6 for checking that f(X) was committed to
+        // correctly.
+        let x_6: C::Scalar = get_challenge_scalar(Challenge(transcript.squeeze().get_lower_128()));
+
+        for eval in self.q_evals.iter() {
+            transcript_scalar.absorb(*eval);
+        }
+
+        let transcript_scalar_point =
+            C::Base::from_bytes(&(transcript_scalar.squeeze()).to_bytes()).unwrap();
+        transcript.absorb(transcript_scalar_point);
+
+        // lagrange_Interpolate polynomial for evaluations at each set
+        let mut r_evals = vec![C::Scalar::zero(); point_sets.len()];
```

Comment:
This implementation can look like the corresponding `fold` improvements made in the prover.